sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,502 @@
+
+### Implementing diagnostics in the DPC++ front-end
+
+As noted above, the front-end device compiler must behave as though there is an
+implicit `[[sycl::requires(has(aspect::fp64))]]` attribute on any use of the
+`double` type in device code.
+
+Aside from this, the front-end compiler can implement the required diagnostics
+purely from the C++ attributes that have been added to the DPC++ headers.
+There is no need for the front-end compiler to know which device features are
+optional.
+
+When the front-end compiler sees a kernel or device function that is decorated
+with `[[sycl::requires()]]`, it forms the set of allowed aspects for that
+kernel or device function using aspects listed in the attribute.  Let's call
+this the `Allowed` set.  The front-end then computes the static call tree of
+that kernel or device function and forms the union of all aspects in any
+`[[sycl::requires()]]` attributes that decorate any of these functions or any
+of the types used inside these functions.  Let's call this the `Used` set.  If
+the `Used` set contains any aspects not in the `Allowed` set, the front-end
+issues a diagnostic.
+
+In order to be user-friendly, the diagnostic should point the user to the
+location of the problem.  Therefore, the diagnostic message should include the
+following information:
+
+* The source position of the `[[sycl::requires()]]` attribute that decorates
+  the kernel or device function which provides the `Allowed` aspect set.  This
+  tells the user which aspects the kernel or device function intends to use.
+
+* The source position of the call to a function that is decorated with
+  `[[sycl::requires()]]` or the source position of the use of a type that is
+  decorated with `[[sycl::requires()]]`.  This tells the user where in the
+  kernel a particular aspect is actually used.
+
+Note that this analysis can be done in the front-end compiler even when a
+kernel makes a call to a function that is in another translation unit.
+Language rules require the application to declare such a function with
+`SYCL_EXTERNAL` in the calling TU, and the `SYCL_EXTERNAL` declaration must be
+decorated with the `[[sycl::requires()]]` attribute.  Therefore, the front-end
+can diagnose errors with aspect usage even without seeing the definition of the
+`SYCL_EXTERNAL` function.
+
+
+## Design to raise required exceptions (and avoid forbidden errors)
+
+As described above the runtime must raise an `errc::kernel_not_supported`
+exception when a kernel is submitted to a device that does not support the
+optional features that the kernel uses.  Likewise, the runtime must **not**
+raise an exception (or otherwise produce an error) due to speculative
+compilation of a kernel for a device, unless the application actually submits
+the kernel to that device.  The solution is largely the same for both JIT and
+AOT cases.
+
+### JIT case
+
+The JIT case requires some change to the way kernels are bundled together into
```

Comment:
In some scenarios binary image may contain no kernels, only external functions (e.g. a dynamic library), but problems outlined below, as well some of as the diagnostic rules still apply. So I suggest to replace 'kernels' with 'functions' where it is applicable. Like here.

Reply:
Good point, thanks.  I updated the design to include details about how this will work when we have exported device functions in shared libraries.  See cda2d8ff750acc1142bb9ab21bd466ed4d23781a.